cmd/snap-bootstrap/partition/sfdisk.go

```diff
@@ -275,12 +323,25 @@ func buildPartitionList(ptable *sfdiskPartitionTable, pv *gadget.LaidOutVolume)
 			s.Label = ubuntuDataLabel
 		}
 
-		toBeCreated = append(toBeCreated, DeviceStructure{p, node})
+		toBeCreated = append(toBeCreated, DeviceStructure{p, node, true})
 	}
 
 	return buf, toBeCreated
 }
 
+// listCreatedPartitions returns a list of indexes of partitions that are
+// marked as created by the installer.
+func listCreatedPartitions(dl *DeviceLayout) []string {
```

Comment:
instead of adding the flag and this, can't we teach buildPartionList to check the for the attribute and return a list of to-be-removed-first partitions based on seen there and the attr instead ?

Reply:
this is also slightly more conservative in the sense that it will remove only things that we consider to create

Reply:
The original reason to have the flag was to bring the attribute information from partition to DeviceStructure in the layout compatibility check which is performed on structures, however it is possible to retrieve this information directly from the partition table data to avoid using the flag. I refactored the code to do that (which resulted in some convergence with the data partition grow PR).

The verification in buildPartitionList is harder to implement because we switched to `sfdisk --append` in 7f92d310 to avoid creating a new partition table and change UUIDs. New partitions can only be added after all existing partitions, so it seems safer to delete all created partitions to always recreate on contiguous space at the end.

Also fixed a test in the process and added extra unit tests for partition removal.

Reply:
To clarify, the order of partitions on disk doesn't need to match the order in the partition table, and the node numbering can be reordered with `sfdisk --reorder`, but in the typical case we have only two or three partitions created at install time and I'm not sure if it's worth the extra complexity.

Reply:
to be clear I'm not asking to have a single sfdisk invocation, I'm wondering if RemoveCreated can't be merged into CreateMissing so that we remove only things that we know will be re-created

Reply:
mmh, of course my proposal is problematic the actual re-install case, let's talk on Monday

Reply:
I think this could work if we also sanity-check partitions marked with the attribute in ensureLayoutCompatibility (an exception for the size of the growable data partition would be needed later). If there's a created partition we don't know about, it would fail at this point and installation won't proceed.